### ISSUE TITLE
Add some way to determine if pairing is necessary

### ISSUE THREAD

**Issue by postlund** [postlund/pyatv#1309]
### What to investigate?

The idea of the public interface in pyatv is that a developer should not have to care about what kind of device is available on the other end. Most things in pyatv already handles this but there's one particular case living in the shadows: pairing. Is pairing needed or not and what credentials to use?

In Home Assistant, I implemented a generic config flow to perform pairing according to how `interface.PairingHandler` works. This flow will automatically handle pairing for all protocols discovered during set up. There's however a new case with the HomePod which it doesn't handle: transient pairing. Transient pairing means that pairing is not needed, but it is internally implemented as a special kind of credentials.

Another case is RAOP, where some devices require pairing, e.g. the Apple TV, but other devices does not, e.g. 3rd party speakers. AirPlay 2 receivers always require at least transient pairing. So I have a few cases to deal with in order to create a new mechanism to tell if pairing is needed or not.

I will try to create a summary of all cases I can think of here (protocols not listed for a device is not supported by the device). Do note that I might have something wrong, this table is based on my observations and some guesses. I've also focused on what's important in context of pyatv, so not all possible pairing methods are necessarily listed.

| Device | Protocol | Auth/Pairing | Comment | How to detect if needed |
| ------- | -------- | ------------ | ---------- | ------------------------ |
| Apple TV 1-3 | AirPlay | Password, None, Legacy | - Password is optional<br /><br />- Legacy authentication is only required if device verification is turned on (otherwise no auth or pairing is needed).| Password:  `(sf & 0x80) or pw`)<br />Device verification (=pairing): `(sf & 0x200)`<br />Reference: https://github.com/postlund/pyatv/issues/1138#issuecomment-863971174
| | DMAP | HSGID, Pairing GUID |  - HSGID is distributed via `hG` in the `_appletv-v2._tcp.local` Zeroconf service. No pairing needed in this case as `hG` can be used as credentials.<br /><br />- Pairing GUID is obtained via DMAP pairing (implemented and works). Always supported. Not necessary if `hG` is present but recommended as multiple clients using `hG` create problems.| HSGID: No pairing<br />Pairing GUID (=pairing): always supported
| | RAOP | Password, None, Legacy | Same as for AirPlay
| Music App (macOS) | DMAP | HSGID | It is possible to interact with the Music app in macOS. It is discovered via the `_hscp._tcp.local`.| Same as AirPlay for Apple TV 1-3 (except for password)
| Apple TV >= 4 | AirPlay | Password, Legacy, HAP | - Password is optional<br /><br />- Either legacy or HAP based authentication is mandatory (prefer HAP if both are supported) and both imply pairing | Password: `(flags & 0x80)`<br /> Legacy: `SupportsLegacyPairing or (sf & 0x200)`<br />HAP: `SupportsCoreUtilsPairingAndEncryption`
| | Companion | HAP | Requires HAP based pairing always (supported and implemented) | Always supported
| | MRP | None, HAP | * No pairing or encryption has ever been needed, was this on purpose?<br /><br />* HAP based pairing | HAP: if `AllowPairing` is `YES` 
| | RAOP | Password, Legacy, HAP | * Same as AirPlay<br /><br />* HAP based pairing requires encryption(?) which legacy does not, so it's easier to use legacy credentials here to avoid having to implement encryption (for now) | Same as AirPlay but `ft` instead of `features` for feature flags
| HomePod | AirPlay | Password, Transient | * Password is optional<br /><br />* Transient pairing (end enabling of encryption) must always be performed. PIN code is hardcoded to 3939 so no actual pairing or device specific credentials are needed, so no "true pairing" is needed | Password: `(flags & 0x80)`<br />Transient: model
| | Companion | HAP | * Companion works as long as credentials are provided (can be extracted from macOS keychain for instance), but no known way to pair with HomePod is known. So is considered not possible for now.| Model (always present via RAOP and/or AirPlay as Companion has no unique identifier)
| | MRP | - |  MRP is tunneled over AirPlay on the HomePod, so pairing is taken care by AirPlay. | Not listed as protocol when scanning, so not an issue
| | RAOP | Password, None, Transient | * Password is optional<br /><br />* No particular encryption is needed<br /><br />* Transient pairing can optionally be used (requires encryption afterwards) | Password: `(flags & 0x80)`<br />Transient: based on model
| AirPort Express | AirPlay | Password, MFi, Transient | * Password is optional<br /><br />* Requires that MFi authentication is initiated, but not necessarily completed (`/auth-setup`). No particular pairing or credentials needed for this. Model (zeroconf property `am`) is compared to `AirPort` to only do this with AirPort Express's as other devices might not like this. | Password: `(sf & 0x80)`<br />MFi: model<br />Transient: model
| | RAOP | Password, Password, MFi, Transient | Same as AirPlay | Same as AirPlay
| Generic AirPlay v1 | AirPlay | Password, None | Most generic AirPlay receivers require no pairing or encryption, but supports password. There are other authentication types not listed here. | Password: `(sf & 0x80) or pw`
| | RAOP | Password, None | Same as AirPlay | Password: `(sf & 0x80)`
| Generic AirPlay v2 | AirPlay | Password, Transient | * Password it optional<br /><br />* Transient pairing is always required (and thus also encryption)<br /><br /> NOTE: Some receivers support both AirPlay 1 and 2, so AirPlay 1 features can be used. | Password: `(flags & 0x80)`<br />Transient: model
| | RAOP | Password, None, Transient | Same as AirPlay | Password: `(sf & 0x80)`<br />

There are a few cases I need to consider how to support. I also need to come up the API part exposing if pairing is needed or not. One suggestion is this:

Extend `interface.BaseService` with a `pairing_state` property with the following type:

```python
class PairingState(Enum):
    NotNeeded = auto()
    """Pairing is not needed, e.g. credentials are not required or credentials already present."""

    Unsupported = auto()
    """Pairing is not supported for this service."""

    Optional = auto()
    """Pairing is supported but not necessary."""

    Required = auto()
    """Pairing must be performed."""
```

Each service/protocol can provide information regarding what kind of pairing is needed. This property shall reflect whether the current service is ready to use, i.e. if all credentials needed are present. One problem here though is that the logic to verify pairing state ends up in the `conf` module and I would prefer to keep it in close to the protocol implementations. So I feel this is not the best approach.

Another way is to add an additional method used to get the state, e.g. `pyatv.pair_state` and let each protocol provide an implementation for the various services. Also border-line OK.

The final solution I'm thinking about is to add the `pairing_state` property (maybe just named `state`) to `interface.PairingHandler` instead. This would require a call to `pyatv.pair` to always be made, but it keeps a very simple API and allows me to keep logic inside the protocol implementation. It would roughly look like:

```python
atvs = await scan(...)
handler = await pair(atvs[0].get_service(Protocol.MRP))
if handler.pairing_state == PairingStateRequired:
    # do pairing
else:
    # something else
```

I would like to include the state in the output after a scan, so I need to figure out how to do that when the information is not part of service object.

### Expected outcome

An API to programmatically know if pairing is needed for a protocol or not.

This is a prerequisite to #1138.

**Comment by postlund** [comment]
I did come up with two issues in my proposed solution:

* Every time `pair` is called, a new client session (aiohttp) is allocated, which is unnecessary
* The `PairingHandler` only gets a handle to the service, but in some cases information from other services are needed indirectly, e.g. device model (Companion is not supported on HomePod, but that can't be derived from the Companion service itself)

So I can't implement it like that. This means I need to discover all services *before* evaluating the pairing state. It makes more sense to provide a generic pairing state for the protocol that does not take available credentials into account and add a helper method that extracts protocols that require pairing. It also makes sense that pairing requirements are discovered during scanning and not as a separate step. That is at least easier to use from an API perspective.

Another thing to consider is that I only do passive detection of pairing requirements right now, i.e. I don't make any outbound connections but only look at zeroconf properties. Maybe at some point I need/want to connect to a protocol to somehow verify if pairing is needed, so I need to make sure that is possible to implement.

My biggest concern right now is how to implement this. I would prefer to keep pairing state with its corresponding service, so that each service keeps track of its pairing state. I end up in a catch-22 as I need all the services of a device to extract the pairing state, but each service shall also contain the pairing state. I would prefer to have a complete `conf.AppleTV` to pass around as that means device information has been extracted, making a lot of things easier to implement.  This suggests that I either have to update the state of each service after the service has been created or not keep state part of services and extract them separately, possibly by another method.

I don't see any reasonable way to untangle what I described above, so I just thought of another way: let the `conf.AppleTV` instance keep track of pairing state for protocols. I can let each protocol define a handler function that extracts pairing state based on services, device information and potentially device address for the future case, where connecting to a device is necessary (will be left out now though). So a handler would look like this:

```python
async def pairing_state(
    device_info: interface.DeviceInfo,
    services: Dict[const.Protocol, interface.BaseServices]
) -> const.PairingState:
    pass
```

Each protocol would implement that function and they would be registered together with scan handlers, so the scan module can extract pairing state after scanning has finished (when all services and device info exists) but before the `const.AppleTV` instance is created. Now we can just add a simple getter in `conf.AppleTV`.

I've used "pairing state" conclusively, but I don't think that is the correct notion here and I will change to "pairing condition" instead, .e.g `const.PairingCondition`.

**Comment by postlund** [comment]
I have managed to implement the general API support now. Next step is to implement support in all protocols according to the table above. Should not be that hard, but there's a lot of work regardless.

**Comment by Schlaubischlump** [comment]
Hi @postlund, 

I quickly looked over the table and I think: Device verification (=pairing): (sf & 0x200) should be for Apple TV >= 4 according to my experiments. Apple TV 1-3 do not require any kind of pairing to use RAOP.

**Comment by postlund** [comment]
@Schlaubischlump So device pairing is only needed for AirPlay, not RAOP at all, on ATV 1-3 if enabled?

**Comment by Schlaubischlump** [comment]
Does the Apple TV 1-3 implement any kind of pairing for AirPlay ? 

I should be able to test if any kind of pairing is required for RAOP or AirPlay. I'll reset my Apple TV 3 to factory settings, not log in with any kind of Apple ID and try to stream some music to it. If some kind of pairing is required, it should show a message.

**Comment by postlund** [comment]
Pairing for ATV 1-3 is device verification (it's technically called "legacy pairing" by apple).

Sounds good 👍

I just drafted a PR with initial work to implement what is described here. I scrapped my latest design idea once more though and did a bigger re-structure. My previous attempt focused only on pairing, so I forgot about passwords. So the implementation I've made now is much more general. Available in #1316.
